File: app/src/main/java/com/fstyle/structure_android/data/source/remote/api/error/BaseException.java
package com.fstyle.structure_android.data.source.remote.api.error;

import android.support.annotation.Nullable;
import com.fstyle.structure_android.data.source.remote.api.response.ErrorResponse;
import java.io.IOException;
import java.net.SocketTimeoutException;
import java.net.UnknownHostException;
import retrofit2.Response;

/**
 * Created by le.quang.dao on 10/03/2017.
 */

public final class BaseException extends RuntimeException {

    @Type
    private final String type;
    @Nullable
    private Response response;
    @Nullable
    private ErrorResponse errorResponse;

    private BaseException(@Type String type, Throwable cause) {
        super(cause.getMessage(), cause);
        this.type = type;
    }

    private BaseException(@Type String type, @Nullable Response response) {
        this.type = type;
        this.response = response;
    }

    public BaseException(@Type String type, @Nullable ErrorResponse response) {
        this.type = type;
        this.errorResponse = response;
    }

    public static BaseException toNetworkError(Throwable cause) {
        return new BaseException(Type.NETWORK, cause);
    }

    public static BaseException toHttpError(Response response) {
        return new BaseException(Type.HTTP, response);
    }

    public static BaseException toUnexpectedError(Throwable cause) {
        return new BaseException(Type.UNEXPECTED, cause);
    }

    public static BaseException toServerError(ErrorResponse response) {
        return new BaseException(Type.SERVER, response);
    }

    @Type
    public String getErrorType() {
        return type;
    }

    public String getMessage() {
        switch (type) {
            case Type.SERVER:
                if (errorResponse != null) {
                    return errorResponse.getMessage();
                }
                return "";
            case Type.NETWORK:
                return getNetworkErrorMessage(getCause());
            case Type.HTTP:
                if (response != null) {
                    return getHttpErrorMessage(response.code());
                }
                return "Error";
            default:
                return "Error";
        }
    }

    private String getNetworkErrorMessage(Throwable throwable) {
        if (throwable instanceof SocketTimeoutException) {
            return throwable.getMessage();
        }

        if (throwable instanceof UnknownHostException) {
            return throwable.getMessage();
        }

        if (throwable instanceof IOException) {
            return throwable.getMessage();
        }

        return throwable.getMessage();
    }

    private String getHttpErrorMessage(int httpCode) {
        if (httpCode >= 300 && httpCode <= 308) {
            // Redirection
            return "It was transferred to a different URL. I'm sorry for causing you trouble";
        }
        if (httpCode >= 400 && httpCode <= 451) {
            // Client error
            return "An error occurred on the application side. Please try again later!";
        }
        if (httpCode >= 500 && httpCode <= 511) {
            // Server error
            return "A server error occurred. Please try again later!";
        }

        // Unofficial error
        return "An error occurred. Please try again later!";
    }
}

File: app/src/main/java/com/ccc/nameproject/data/source/remote/api/error/BaseException.kt
package com.ccc.nameproject.data.source.remote.api.error

/**
 * Created by daolq on 5/5/18.
 */
import retrofit2.Response
import java.io.IOException
import java.net.SocketTimeoutException
import java.net.UnknownHostException

/**
 * Created by le.quang.dao on 10/03/2017.
 */

class BaseException : RuntimeException {

  val errorType: String
  private lateinit var response: Response<*>
  private var errorResponse: ErrorResponse? = null

  private constructor(type: String, cause: Throwable) : super(cause.message, cause) {
    this.errorType = type
  }

  private constructor(type: String, response: Response<*>) {
    this.errorType = type
    this.response = response
  }

  constructor(type: String, response: ErrorResponse?) {
    this.errorType = type
    this.errorResponse = response
  }

  fun getMessageError(): String {
    when (errorType) {
      Type.SERVER -> {
        errorResponse?.error?.message.let {
          return it!!
        }
      }
      Type.NETWORK -> return getNetworkErrorMessage(cause)
      Type.HTTP -> {
        return response.code().getHttpErrorMessage()
      }
      else -> return "Error"
    }
  }

  private fun getNetworkErrorMessage(throwable: Throwable?): String {
    if (throwable is SocketTimeoutException) {
      return throwable.message.toString()
    }

    if (throwable is UnknownHostException) {
      return throwable.message.toString()
    }

    if (throwable is IOException) {
      return throwable.message.toString()
    }

    return throwable?.message.toString()
  }

  private fun Int.getHttpErrorMessage(): String {
    if (this in 300..308) {
      // Redirection
      return "It was transferred to a different URL. I'm sorry for causing you trouble"
    }
    if (this in 400..451) {
      // Client error
      return "An error occurred on the application side. Please try again later!"
    }
    if (this in 500..511) {
      // Server error
      return "A server error occurred. Please try again later!"
    }

    // Unofficial error
    return "An error occurred. Please try again later!"
  }

  companion object {

    fun toNetworkError(cause: Throwable): BaseException {
      return BaseException(Type.NETWORK, cause)
    }

    fun toHttpError(response: Response<*>): BaseException {
      return BaseException(Type.HTTP, response)
    }

    fun toUnexpectedError(cause: Throwable): BaseException {
      return BaseException(Type.UNEXPECTED, cause)
    }

    fun toServerError(response: ErrorResponse): BaseException {
      return BaseException(Type.SERVER, response)
    }
  }
}